### DISCUSSION TITLE
No Validation inside cellEditor?

### DISCUSSION THREAD

**Discussion by DerIngo** [primefaces/community#1766] [PrimeFaces] [answered]
I have this simple DataTable:
```      <h:form id="frmTest">
        <p:dataTable value="#{testView.list}" var="item" editable="true" editMode="cell">
          <p:column width="100" headerText="Name">
            <p:cellEditor>
              <f:facet name="output">
                <h:outputText value="Name: #{item.name}" />
              </f:facet>
              <f:facet name="input">
                <p:inputText value="#{item.name}" required="true"></p:inputText>
              </f:facet>
            </p:cellEditor>
          </p:column>
        </p:dataTable>
        <p:commandButton value="Test" action="#{testView.action}" />
      </h:form>
```
Even when item.name is NULL, the commandButton executes testView.action.
I expected validation to fail BEFORE the action method of the commandButton is executed.

Here is my [primefaces-test](https://github.com/DerIngo/primefaces-test) fork.
You can execute the sample with <strong>mvn jetty:run</strong> command and hit <strong>http://localhost:8080/</strong> to run the page.

**Comment by melloware** [comment]
I think this has been asked before.  When you click the button its executing first. To get in front of it add a `blur` event that updates the `@form` to your input like this.

```xml
<p:inputText value="#{item.name}" required="true"><p:ajax event="blur" process="@this" update="@form"/></p:inputText>
```

**Comment by DerIngo** [comment]
I have added the `blur` event as suggested, but it makes no difference: I can still press the button and the action is executed.

I assume the problem is something like:
The page is loaded and displays all the `<f:facet name="output">`-blocks.
So no `<f:facet name="input">`-block is in the DOM when the button is clicked and so there is no validation of this "not existing" inputs.

**Reply by melloware** [reply]
Correct.

**Reply by DerIngo** [reply]
So i would consider this as a bug. And a bug should be fixed.
But I am afraid, the fix would be very complex (with possible sideeffects?). And it seems, that nearly no one else has this problem, which really surprises me. So I guess, this issue will not get high priority the next years.

Any better idea for a workaraound, than validating in the action method?
```
    public String action() {
        FacesContext context = FacesContext.getCurrentInstance();
        boolean validationFailed = false;
        
        // Check if any inputText in the dataTable has invalid value
        for (TestObject testObject : list) {
            if (testObject.getName() == null || testObject.getName().isEmpty()) {
                context.addMessage(null, new FacesMessage(FacesMessage.SEVERITY_ERROR, "Validation Error", "Name is required."));
                validationFailed = true;
            }
        }
        
        // If validation failed, return null to stay on the same page
        if (validationFailed) {
            return null;
        }        
        
        // validation passed, go on with business logic

    }
```

**Reply by melloware** [reply]
Yes CellEditor has always been quite odd.  I think most people have chosen not to use it and instead use the Row Editing feature instead which seems to have much less issues because it forces you to "commit" the row before moving on.

**Reply by bdonauba** [reply]
I just stumbled across this limitation. I guess it is the same reason why my bean validation does not  work along with <p:datatable> and cell edit mode. 

I have several NotEmpty annotated fields in my entity beans. If I don't fill this input fields I normally get a validation error with messages in <p:message>. In case of <p:cellEditor> I do not get any validation error in the "process validation" phase, but later when JPA tries to persist the entity it throws an exception that is shown in the browser and logs.

I think this behaviour should be clearly and bold stated in <p:datatable> documentation. It took my some time till I found this post making it clear that this is a primefaces limitation not a error in my code.

**Reply by melloware** [reply]
@bdonauba @DerIngo   a PR to update the documentation is welcome!

**Reply by bdonauba** [reply, marked as answer]
Fair enough.

In regard to my statement from November the 20 concerning bean validation I have to say it was an assumption of mine. I never checked the code but I think you are an employee of primetek so I guess my suspicion is true.

I would add a warning here https://github.com/primefaces/primefaces/blob/master/docs/16_0_0/components/datatable.md#editing. Probably all other versions of primefaces are also concerned.

**Reply by melloware** [reply]
@bdonauba i don't work for PrimeTek I am just a user just like you...  So PR for documentation update is welcome!
